druid-shell/src/platform/x11/window.rs

```diff
@@ -401,30 +355,23 @@ impl Window {
     }
 
     fn render(&self, invalid_rect: Rect) -> Result<(), Error> {
-        // TODO(x11/errors): this function should return a an error
-        // instead of panicking or logging if the error isn't recoverable.
-
         // Figure out the window's current size
         let geometry_cookie = xcb::get_geometry(self.app.connection(), self.id);
-        let reply = geometry_cookie.get_reply().unwrap();
+        let reply = geometry_cookie.get_reply()?;
         let size = Size::new(reply.width() as f64, reply.height() as f64);
         self.set_size(size)?;
 
-        let mut anim = false;
-        if let Ok(mut cairo_ctx) = self.cairo_context.try_borrow_mut() {
+        let anim;
+        {
+            let mut cairo_ctx = borrow_mut!(self.cairo_context)?;
             let mut piet_ctx = Piet::new(&mut cairo_ctx);
             piet_ctx.clip(invalid_rect);
-            if let Ok(mut handler) = self.handler.try_borrow_mut() {
-                anim = handler.paint(&mut piet_ctx, invalid_rect);
-            } else {
-                log::error!("Window::render - handler already borrowed");
-            }
-            if let Err(e) = piet_ctx.finish() {
-                log::error!("Window::render - piet finish failed: {}", e);
-            }
+
+            anim = borrow_mut!(self.handler)?.paint(&mut piet_ctx, invalid_rect);
```

Comment:
We can't return an error here this easily. This will crash the program, because `piet_ctx` needs to be finished before it's dropped.

Probably also need to reset the cairo clip even if finishing fails.

I think it still makes sense to return the error from the function, however we can't do it immediately and need to try to perform all the cleanup we can.

Reply:
Ok, good point. I've redone this part.